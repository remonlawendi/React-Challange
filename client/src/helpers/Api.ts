import Axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, CancelTokenSource } from "axios";
import ICase from "../models/ICase";
import IFeature from "../models/IFeature";

const config: AxiosRequestConfig = {
    baseURL: "https://bikewise.org:443/api/v2/",
    responseType: "json",
};

const axios: AxiosInstance = Axios.create(config);
export const fetchSourceAll: CancelTokenSource = Axios.CancelToken.source();
export const fetchSourceDetails: CancelTokenSource = Axios.CancelToken.source();


interface ICasesResponse {
    incidents: ICase[];
}

interface ApiResponse<T> extends AxiosResponse<T> {
    data: T;
    status: number;
    statusText: string;
    headers: any;
    config: AxiosRequestConfig;
}
/**
 * Description Get all theft cases from Bikewise API, it's not paginated due to API limitations
 */
export const fetchAllCases = async (search?: string, from?: string, to?: string) => {
   const toTimeStamp = (date?: string) => date ? Math.ceil(Date.parse(date) / 1000) : "";
   const theftCases: ApiResponse<ICasesResponse> = await axios.request<ICasesResponse>({
     url: "/incidents",
     method: "get",
     params: { proximity: "berlin", incident_type: "theft", query: search, occurred_before: toTimeStamp(to), occurred_after: toTimeStamp(from) },
     cancelToken: fetchSourceAll.token,
   });
   return theftCases.data.incidents;
};
interface ICaseResponse {
    incident: ICase;
}
interface ILocationsResponse {
     features: IFeature[];
}

/**
 * Description get and return theft case details
 */
export const fetchCaseDetails = async (id: string) => {
  const theftCase: ApiResponse<ICaseResponse> = await axios.request<ICaseResponse>({
     url: `/incidents/${id}`,
     method: "get",
     params: { proximity: "berlin", incident_type: "theft" },
     cancelToken: fetchSourceDetails.token,
   });
  const { incident } = theftCase.data;
  const features: ApiResponse<ILocationsResponse> = await axios.request<ILocationsResponse>({
     url: "/locations",
     method: "get",
     params: { query: incident.title, proximity: "berlin", incident_type: "theft" },
     cancelToken: fetchSourceDetails.token,
   });
  incident.feature = features.data.features.find((f) => f.properties.id === incident.id);
  return incident;
};
