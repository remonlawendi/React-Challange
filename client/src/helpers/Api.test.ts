import mockAxios from "jest-mock-axios";
import { HttpResponse } from "jest-mock-axios/dist/lib/mock-axios-types";
import { fetchAllCases } from "./Api";

it("FetchAllCases should return correct data", async () => {
  const doFetch = fetchAllCases();
  const response: HttpResponse = {
    data: {
        incidents: [
        {
          id: 102152,
          title: "Post to Neighbors",
          description: "wondering if all the birch trees",
          address: "801 Van White Memorial Blvd Minneapolis, MN 55411, USA",
          occurred_at: 1560614519,
          updated_at: 1560620028,
          url: "https://bikewise.org/api/v1/incidents/102152",
          source: {
            name: "SeeClickFix.com",
            html_url: "https://seeclickfix.com/issues/6034466",
            api_url: "https://seeclickfix.com/api/v2/issues/6034466",
          },
        }],
    },
    status: 200,
  };
  mockAxios.mockResponse(response);
  const result = await doFetch;
  expect(result).toMatchSnapshot();
});
