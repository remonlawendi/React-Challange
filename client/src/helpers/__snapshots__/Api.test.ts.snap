// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchAllCases should return correct data 1`] = `
Array [
  Object {
    "address": "801 Van White Memorial Blvd Minneapolis, MN 55411, USA",
    "description": "wondering if all the birch trees",
    "id": 102152,
    "occurred_at": 1560614519,
    "source": Object {
      "api_url": "https://seeclickfix.com/api/v2/issues/6034466",
      "html_url": "https://seeclickfix.com/issues/6034466",
      "name": "SeeClickFix.com",
    },
    "title": "Post to Neighbors",
    "updated_at": 1560620028,
    "url": "https://bikewise.org/api/v1/incidents/102152",
  },
]
`;
